### PR TITLE
[Resolver] Modify sorting heuristic to check how constraining the req…

### PR DESCRIPTION
**Pull request by segiddins**
…uirements are

This should speed up cases where a simple-looking requirement actually eliminates a lot of the possibilities

\c @indirect @Who828

### CODE REVIEW

File: lib/bundler/resolver.rb [hunk 1]
```diff
@@ -296,14 +295,23 @@ def sort_dependencies(dependencies, activated, conflicts)
...
           activated.vertex_named(name).payload ? 0 : search_for(dependency).count,
         ]
       end
     end
 
+    def amount_constrained(dependency)
+      @amount_constrained ||= {}
+      @amount_constrained[dependency.name] ||= begin
+        base_dep = Dependency.new dependency.name, '>= 0.a'
+        all = search_for(DepProxy.new base_dep, dependency.__platform)
```
Comment (indirect) [inline lib/bundler/resolver.rb]: Doesn't calling __platform cause a request to download the gemspec? I worry that this is really, really slow when we don't already have the platform of the dep...

**Comment by indirect** [comment]
So if I'm understanding this correctly, it's sorting dependencies whose constraints remove the largest percentage of the total search space to the top, right? That sounds like a good deal to me. Do we have  Gemfiles we can performance test this against?

File: lib/bundler/resolver.rb (hunk 1 above)
Comment (segiddins) [inline lib/bundler/resolver.rb]: pretty sure `dependency` here is a `DepProxy` that is initialized with a `platform`

**Comment by segiddins** [comment]
Try this (using `bundle lock --print`) against the `gxapi_rails` realworld spec

**Comment by indirect** [comment]
This doesn't seem to slow down the common cases, and this speeds up some really bad cases by a huge amount. :+1:

**Comment by segiddins** [comment]
Fwiw this sped up https://github.com/bundler/bundler/issues/3634 to the point where that Gemfile takes under a second and a half to `bundle lock --print`

**Comment by indirect** [comment]
Awesome.

**Pull request merged**